Zend/zend_language_scanner.l

```diff
@@ -2104,21 +2158,97 @@ inline_char_handler:
 	}
 
 	heredoc_label->label = estrndup(s, heredoc_label->length);
+	heredoc_label->indentation = -1;
+	saved_cursor = YYCURSOR;
+
+	zend_ptr_stack_push(&SCNG(heredoc_label_stack), (void *) heredoc_label);
+
+	while (YYCURSOR < YYLIMIT && (*YYCURSOR == ' ' || *YYCURSOR == '\t')) {
+		spacing |= 1 << (*YYCURSOR == '\t');
+		++YYCURSOR;
+	}
+
+	if (YYCURSOR == YYLIMIT) {
+		YYCURSOR = saved_cursor;
+		RETURN_TOKEN(T_START_HEREDOC);
+	}
 
 	/* Check for ending label on the next line */
 	if (heredoc_label->length < YYLIMIT - YYCURSOR && !memcmp(YYCURSOR, s, heredoc_label->length)) {
 		YYCTYPE *end = YYCURSOR + heredoc_label->length;
 
-		if (*end == ';') {
-			end++;
-		}
+		if (!IS_LABEL_START(*end)) {
+			if ((spacing & 3) == 3) {
```

Comment:
What is this magic number 3?

Reply:
I could change this to be:
```c
if (spacing & (USES_TABS | USES_SPACES)) {
```

The `while` loop a few lines above should probably also be changed then to:
```c
while (YYCURSOR < YYLIMIT && (*YYCURSOR == ' ' || *YYCURSOR == '\t')) {
	spacing |= *YYCURSOR == '\t' ? USES_TABS : USES_SPACES;
```